src/libcore/ffi.rs

```diff
@@ -369,5 +420,96 @@ extern "rust-intrinsic" {
     /// Loads an argument of type `T` from the `va_list` `ap` and increment the
     /// argument `ap` points to.
     #[cfg(not(bootstrap))]
+    #[cfg(not(all(target_arch = "aarch64", not(target_os = "ios"), not(windows))))]
     fn va_arg<T: sealed_trait::VaArgSafe>(ap: &mut VaListImpl<'_>) -> T;
 }
+
+#[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
+use crate::{mem, ptr};
+
+// Implemenation for AArch64 linux (and others that follow the standard PCS)
+// based on https://developer.arm.com/docs/ihi0055/d/procedure-call-standard-for-the-arm-64-bit-architecture
+// APPENDIX Variable argument Lists -> The va_start() macro
+#[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
+impl<'f> VaListImpl<'f> {
+    unsafe fn va_arg_on_stack<T: sealed_trait::VaArgSafe>(&mut self) -> T {
+        let mut arg = self.stack as usize;
+        if mem::align_of::<T>() > 8 {
+            arg = (arg + 15) & !15;
+        }
+        self.stack = ((arg + mem::size_of::<T>() + 7) & !7) as *mut c_void;
+        #[cfg(target_endian = "big")]
+        {
+            if
+            /* classof(type) != "aggregate" && */
+            mem::size_of::<T>() < 8 {
+                arg += 8 - mem::size_of::<T>();
+            }
+        }
+        ptr::read(arg as *const T)
+    }
+
+    unsafe fn va_arg_gr<T: sealed_trait::VaArgSafe>(&mut self) -> T {
+        let mut offs = self.gr_offs;
+        if offs >= 0 {
+            return self.va_arg_on_stack(); // reg save area empty
+        }
+        if mem::align_of::<T>() > 8 {
+            offs = (offs + 15) & !15; // round up
+        }
+        let nreg = (mem::size_of::<T>() as i32 + 7) / 8;
+        self.gr_offs = offs + (nreg * 8);
+        if self.gr_offs > 0 {
+            return self.va_arg_on_stack(); // overflowed reg save area
```

Comment:
This reminds me I wanted to change how `rustc_target::abi::call` works for a long time, and make it compute "designations" such as "on-stack" or "in register, if available, otherwise on-stack", with per-target "resources" (such as general-purpose registers for args/return and floating-point registers etc.), and put all of that together to compute final "assignments", *afterwards* (and we could even cache the per-type computation).
If we also tracked *which* registers were being used, that might even help Cranelift (cc @sunfishcode).

The relevant point here is making the per-type information context-free, so that it can be used for both signatures *and* `va_arg`.

Reply:
That sounds ideal but outside my current knowledge/ability. Perhaps I could have a look at that after fixing it this way.